src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorNodeWriter.cs

```diff
@@ -49,5 +53,169 @@ public sealed override void WriteCSharpCodeAttributeValue(CodeRenderingContext c
             return;
         }
+
+
+        // Currently the same for design time and runtime
+        public void WriteComponentTypeInferenceMethod(CodeRenderingContext context, ComponentTypeInferenceMethodIntermediateNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // This is ugly because CodeWriter doesn't allow us to erase, but we need to comma-delimit. So we have to
+            // materizalize something can iterate, or use string.Join. We'll need this multiple times, so materializing
+            // it.
+            var parameters = GetParameterDeclarations();
+
+            // This is really similar to the code in WriteComponentAttribute and WriteComponentChildContent - except simpler because
+            // attributes and child contents look like variables. 
+            //
+            // Looks like:
+            //
+            //  public static void CreateFoo_0<T1, T2>(RenderTreeBuilder builder, int seq, int __seq0, T1 __arg0, int __seq1, global::System.Collections.Generic.List<T2> __arg1, int __seq2, string __arg2)
+            //  {
+            //      builder.OpenComponent<Foo<T1, T2>>();
+            //      builder.AddAttribute(__seq0, "Attr0", __arg0);
+            //      builder.AddAttribute(__seq1, "Attr1", __arg1);
+            //      builder.AddAttribute(__seq2, "Attr2", __arg2);
+            //      builder.CloseComponent();
+            //  }
+            //
+            // As a special case, we need to generate a thunk for captures in this block instead of using
+            // them verbatim.
+            //
+            // The problem is that RenderTreeBuilder wants an Action<object>. The caller can't write the type
+            // name if it contains generics, and we can't write the variable they want to assign to. 
+            var writer = context.CodeWriter;
+
+            writer.Write("public static void ");
+            writer.Write(node.MethodName);
+
+            writer.Write("<");
+            writer.Write(string.Join(", ", node.Component.Component.GetTypeParameters().Select(a => a.Name)));
+            writer.Write(">");
+
+            writer.Write("(");
+            writer.Write("global::");
+            writer.Write(BlazorApi.RenderTreeBuilder.FullTypeName);
+            writer.Write(" builder");
+            writer.Write(", ");
+            writer.Write("int seq");
+
+            if (parameters.Count > 0)
+            {
+                writer.Write(", ");
+            }
+
+            for (var i = 0; i < parameters.Count; i++)
+            {
+                writer.Write("int ");
+                writer.Write(parameters[i].seqName);
+
+                if (parameters[i].typeName != null && parameters[i].parameterName != null)
```

Comment:
Could you explain these conditions a bit more? It looks to me like `parameterName` is certainly always nonnull due to how `GetParameters` is implemented, and I don't know what scenario would be represented by `typeName` being null.

Reply:
yeah, sorry this was cleanup I missed when I was iterating. I originally started with rendering the 'ref' lambdas inside the generated thunk, but it doesn't really work.